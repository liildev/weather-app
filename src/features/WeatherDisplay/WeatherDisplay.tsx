import type React from 'react';
import type { WeatherResponse } from '@/shared/api/types';
import {
	convertTemperature,
	getTemperatureUnit,
	getWeatherIcon,
} from '@/shared/lib/utils';
import { LoadingSpinner } from '@/shared/ui/LoadingSpinner';

interface WeatherDisplayProps {
	weather: WeatherResponse | null;
	unit: 'celsius' | 'fahrenheit';
	loading: boolean;
}

export const WeatherDisplay: React.FC<WeatherDisplayProps> = ({
	weather,
	unit,
	loading,
}) => {
	if (loading) {
		return (
			<div className='weather-card flex items-center justify-center h-48'>
				<LoadingSpinner size='lg' />
			</div>
		);
	}

	if (!weather) {
		return (
			<div className='weather-card flex items-center justify-center h-48'>
				<p className='text-gray-500 dark:text-gray-400'>
					No weather data available
				</p>
			</div>
		);
	}

	const temperature = convertTemperature(weather.main.temp, unit);
	const feelsLike = convertTemperature(weather.main.feels_like, unit);
	const tempUnit = getTemperatureUnit(unit);

	return (
		<div className='weather-card animate-fade-in'>
			<div className='text-center mb-6'>
				<div className='text-6xl mb-2'>
					{getWeatherIcon(weather.weather[0].icon)}
				</div>
				<h2 className='text-3xl font-bold mb-2'>
					{temperature}
					{tempUnit}
				</h2>
				<p className='text-lg text-gray-600 dark:text-gray-400 capitalize'>
					{weather.weather[0].description}
				</p>
				<p className='text-sm text-gray-500 dark:text-gray-500'>
					Feels like {feelsLike}
					{tempUnit}
				</p>
			</div>

			<div className='grid grid-cols-2 gap-4'>
				<div className='text-center p-3 bg-gray-50 dark:bg-gray-700 rounded-lg'>
					<p className='text-sm text-gray-600 dark:text-gray-400'>Humidity</p>
					<p className='text-lg font-semibold'>{weather.main.humidity}%</p>
				</div>
				<div className='text-center p-3 bg-gray-50 dark:bg-gray-700 rounded-lg'>
					<p className='text-sm text-gray-600 dark:text-gray-400'>Pressure</p>
					<p className='text-lg font-semibold'>{weather.main.pressure} hPa</p>
				</div>
			</div>
		</div>
	);
};
