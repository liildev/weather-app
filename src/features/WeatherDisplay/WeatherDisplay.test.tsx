import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { WeatherDisplay } from './WeatherDisplay';
import type { WeatherResponse } from '@/shared/api/types';

const mockWeatherData: WeatherResponse = {
	weather: [{ id: 800, main: 'Clear', description: 'clear sky', icon: '01d' }],
	main: {
		temp: 25,
		feels_like: 27,
		temp_min: 22,
		temp_max: 28,
		pressure: 1013,
		humidity: 60,
	},
	name: 'London',
	dt: Date.now() / 1000,
};

describe('WeatherDisplay', () => {
	it('should render loading spinner when loading', () => {
		const { container } = render(
			<WeatherDisplay weather={null} unit='celsius' loading={true} />
		);

		expect(container.querySelector('.animate-spin')).toBeInTheDocument();
	});

	it('should render no data message when weather is null', () => {
		const { getByText } = render(
			<WeatherDisplay weather={null} unit='celsius' loading={false} />
		);

		expect(getByText('No weather data available')).toBeInTheDocument();
	});

	it('should render weather data correctly', () => {
		const { getByText } = render(
			<WeatherDisplay
				weather={mockWeatherData}
				unit='celsius'
				loading={false}
			/>
		);

		expect(getByText('25째C')).toBeInTheDocument();
		expect(getByText('clear sky')).toBeInTheDocument();
		expect(getByText('Feels like 27째C')).toBeInTheDocument();
		expect(getByText('60%')).toBeInTheDocument();
		expect(getByText('1013 hPa')).toBeInTheDocument();
	});

	it('should convert temperature to fahrenheit', () => {
		const { getByText } = render(
			<WeatherDisplay
				weather={mockWeatherData}
				unit='fahrenheit'
				loading={false}
			/>
		);

		expect(getByText('77째F')).toBeInTheDocument();
		expect(getByText('Feels like 81째F')).toBeInTheDocument();
	});

	it('should match snapshot', () => {
		const { container } = render(
			<WeatherDisplay
				weather={mockWeatherData}
				unit='celsius'
				loading={false}
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});
});
