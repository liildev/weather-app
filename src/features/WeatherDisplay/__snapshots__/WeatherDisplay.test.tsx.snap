// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeatherDisplay > should match snapshot 1`] = `
<div
  class="weather-card animate-fade-in"
>
  <div
    class="text-center mb-6"
  >
    <div
      class="text-6xl mb-2"
    >
      ☀️
    </div>
    <h2
      class="text-3xl font-bold mb-2"
    >
      25
      °C
    </h2>
    <p
      class="text-lg text-gray-600 dark:text-gray-400 capitalize"
    >
      clear sky
    </p>
    <p
      class="text-sm text-gray-500 dark:text-gray-500"
    >
      Feels like 
      27
      °C
    </p>
  </div>
  <div
    class="grid grid-cols-2 gap-4"
  >
    <div
      class="text-center p-3 bg-gray-50 dark:bg-gray-700 rounded-lg"
    >
      <p
        class="text-sm text-gray-600 dark:text-gray-400"
      >
        Humidity
      </p>
      <p
        class="text-lg font-semibold"
      >
        60
        %
      </p>
    </div>
    <div
      class="text-center p-3 bg-gray-50 dark:bg-gray-700 rounded-lg"
    >
      <p
        class="text-sm text-gray-600 dark:text-gray-400"
      >
        Pressure
      </p>
      <p
        class="text-lg font-semibold"
      >
        1013
         hPa
      </p>
    </div>
  </div>
</div>
`;
